pkg/controller/azvolumeattachment.go

```diff
@@ -0,0 +1,546 @@
+	nodeMap    map[string]map[string]v1alpha1.AzVolumeAttachment
+	scMap      map[string]map[string]bool
+	volMap     map[string]string
+}
+
+type filteredNode struct {
+	azDriverNode v1alpha1.AzDriverNode
+	numAttached  int
+}
+
+var _ reconcile.Reconciler = &reconcileAzVolumeAttachment{}
+
+func (r *reconcileAzVolumeAttachment) Reconcile(ctx context.Context, request reconcile.Request) (reconcile.Result, error) {
+	if !r.IsInitialized() {
+		if err := r.SyncStorageClass(ctx); err != nil {
+			return reconcile.Result{Requeue: true}, err
+		}
+		if err := r.SyncVolumeAttachments(ctx); err != nil {
+			return reconcile.Result{Requeue: true}, err
+		}
+		klog.Info("Successfully initialized maps for azVolumeAttachment Controller")
+	}
+	// check if request name involves change in storage class
+	// TODO: need to know what kind of error would be returned if r.client.get is called on a wrong resource type
+	klog.Infof("name: %s, namespace:%s", request.Name, request.Namespace)
+	result, err := r.HandleReplicaUpdate(ctx, request)
+	if err != nil && errors.IsNotFound(err) {
+		// if not assume that an event occurred in azVolumeAttachment object
+		return r.HandleAzVolumeAttachmentEvent(ctx, request)
+	}
+	return result, err
+}
+
+func (r *reconcileAzVolumeAttachment) HandleAzVolumeAttachmentEvent(ctx context.Context, request reconcile.Request) (reconcile.Result, error) {
+	var azVolumeAttachment v1alpha1.AzVolumeAttachment
+	if err := r.client.Get(ctx, request.NamespacedName, &azVolumeAttachment); err != nil {
+		// if the azVolumeAttachment object can no longer be found, it's been deleted so do not requeue
+		if errors.IsNotFound(err) {
+			return reconcile.Result{}, nil
+		}
+		// if the GET failure is triggered by other errors, log it and requeue the request
+		klog.Errorf("failed to fetch azvolumeattachment object with namespaced name %s: (%v)", request.NamespacedName, err)
+		return reconcile.Result{Requeue: true}, err
+	}
+
+	// this is a creation event
+	if azVolumeAttachment.Status == nil {
+		// First, attach the volume to the specified node
+		if err := r.AttachVolume(ctx, azVolumeAttachment.Spec.UnderlyingVolume, azVolumeAttachment.Spec.NodeName); err != nil {
+			klog.Errorf("failed to attach volume %s to node %s: (%v)", azVolumeAttachment.Spec.UnderlyingVolume, azVolumeAttachment.Spec.NodeName, err)
+			return reconcile.Result{Requeue: true}, err
+		}
+
+		// If it is the first time seeing this underlying volume, sync scMap and volMap
+		if _, ok := r.volMap[azVolumeAttachment.Spec.UnderlyingVolume]; !ok {
+			if err := r.SyncStorageClass(ctx); err != nil {
+				return reconcile.Result{Requeue: true}, err
+			}
+		}
+
+		// update the status
+		updated := azVolumeAttachment.DeepCopy()
+		updated.Status = &v1alpha1.AzVolumeAttachmentStatus{}
+		updated.Status.AttachmentTier = updated.Spec.AttachmentTier
+		updated.Status.AttachmentState = v1alpha1.Attached
+
+		if azVolumeAttachment.Spec.AttachmentTier == v1alpha1.Primary {
```

Comment:
It would be useful to break this method up into separate methods for each scenario, e.g. `AttachDisk`, `DetachDisk` and `ChangeDiskAttachementRole`.

Reply:
Done :)